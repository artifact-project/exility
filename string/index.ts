import {requiredScopeKeys} from '@exility/block';
import {core as stdlib} from '@exility/stdlib';
import createCompiler, {StringModeOptions} from './src/compiler/compiler';

function runtimeBlockActivate(Block, options: Partial<StringModeOptions>) {
	if (!Block.prototype['__template__']) {
		const compile = createCompiler({
			...options,
			blocks: Object.keys(Block.blocks),
			scope: requiredScopeKeys,
			cssModule: !!Block.classNames,
		});
		const templateFactory = compile(Block.template);

		Block.prototype['name'] = Block.prototype['name'] || Block.name;
		Block.prototype['__template__'] = templateFactory({stdlib});

		return compile;
	}
}

export {
	createCompiler,
	runtimeBlockActivate,
};
