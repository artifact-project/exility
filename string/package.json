{
  "name": "@exility/strong",
  "version": "0.1.0",
  "description": "Exility Compiler to Pure String",
  "main": "index.ts",
  "scripts": {
    "test": "jest --coverage"
  },
  "author": "RubaXa <trash@rubaxa.org>",
  "license": "MIT",
  "devDependencies": {
    "@types/jest": "^19.2.3",
    "jest": "^20.0.3",
    "ts-jest": "^20.0.4",
    "typescript": "^2.3.3"
  },
  "peerDependencies": {
    "@exility/parser": "*",
    "@exility/compile-helpers": "*"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "testRegex": "\\.tests\\.ts$",
    "transformIgnorePatterns": [
      "node_modules/(?!@exility)"
    ],
    "setupTestFrameworkScriptFile": "<rootDir>/__jest__/extensions.ts",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "json"
    ]
  }
}
