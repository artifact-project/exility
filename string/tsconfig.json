{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "declaration": true,
    "moduleResolution": "node",
    "experimentalDecorators": true
  },
  "exclude": ["dist", "node_modules"]
}
