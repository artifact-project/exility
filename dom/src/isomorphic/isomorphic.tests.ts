import {
	core as stdlib,
	dom as stddom,
} from '@exility/stdlib';
import {createCompiler as createStringCompiler} from '@exility/string';

import createDOMCompiler from '../compiler/compiler';
import {pause} from '../compiler/compiler.blocks.tests';

const stringCompiler = createStringCompiler({
	scope: ['x', 'y', 'z'],
	metaComments: true,
});

const domCompiler = createDOMCompiler({
	scope: ['x', 'y', 'z'],
	isomorphic: true,
});

const stringCompilerWithBlocks = createStringCompiler({
	blocks: ['XFoo'],
	scope: ['x', 'y', 'z', '__blocks__'],
	metaComments: true,
});

const domCompilerWithBlocks = createDOMCompiler({
	blocks: ['XFoo'],
	scope: ['x', 'y', 'z', '__blocks__'],
	isomorphic: true,
});

function fromString(tpl, initialData, actualData, debug?) {
	const stringTemplate = stringCompiler(tpl)({stdlib});
	const domTemplate = domCompiler(tpl)({stdlib, stddom});
	const container = document.createElement('div');

	container.innerHTML = stringTemplate(initialData);
	debug && console.log(domTemplate.toString());

	const view = domTemplate(actualData, {isomorphic: container});

	view.mountTo(container);

	return {view, container};
}

function fromStringWithBlocks(tpl, initialData, actualData, debug?) {
	const stringTemplate = stringCompilerWithBlocks(tpl)({stdlib});
	const domTemplate = domCompilerWithBlocks(tpl)({stdlib, stddom});
	const container = document.createElement('div');

	initialData.__blocks__ = {...initialData.__blocks__};
	container.innerHTML = stringTemplate(initialData);
	debug && console.log(domTemplate.toString());

	const view = domTemplate(actualData, {isomorphic: container});

	view.mountTo(container);

	return {view, container};
}

it('iso / value', () => {
	const {view, container} = fromString(
		'h1 | ${x}?\n.is-${y} | Hi, ${z}, how are you?',
		{z: 'fail'},
		{x: 'Wow!', y: 'user', z: 'Zorro'},
	);

	expect(container.innerHTML).toBe('<h1><!--(-->Wow!<!--)-->?</h1><div class=\"is-user\">Hi, <!--(-->Zorro<!--)-->, how are you?</div>');

	view.update({x: 'ok', y: 'super-user', z: 'RubaXa'});
	expect(container.innerHTML).toBe('<h1><!--(-->ok<!--)-->?</h1><div class=\"is-super-user\">Hi, <!--(-->RubaXa<!--)-->, how are you?</div>');
});

it('iso / if / false -> false', () => {
	const {view, container} = fromString(
		'if (x)\n  | ${y}\nb | ${z}',
		{z: 1},
		{z: 1},
	);

	expect(container.innerHTML).toBe('<!--if1--><!--/if1--><b><!--(-->1<!--)--></b>');

	view.update({x: true, y: 'foo'});
	expect(container.innerHTML).toBe('<!--if1-->foo<!--/if1--><b><!--(--><!--)--></b>');

	view.update({x: true, y: 'bar', z: 11});
	expect(container.innerHTML).toBe('<!--if1-->bar<!--/if1--><b><!--(-->11<!--)--></b>');
});

it('iso / if / false -> true', () => {
	const {view, container} = fromString(
		'if (x)\n  | ${y}\nb | ${z}',
		{z: 2},
		{x: true, y: 'foo', z: 2},
	);

	expect(container.innerHTML).toBe('<!--if1-->foo<!--/if1--><b><!--(-->2<!--)--></b>');

	view.update({x: true, y: 'bar', z: 22});
	expect(container.innerHTML).toBe('<!--if1-->bar<!--/if1--><b><!--(-->22<!--)--></b>');

	view.update({});
	expect(container.innerHTML).toBe('<!--if1--><!--/if1--><b><!--(--><!--)--></b>');
});

it('iso / if / true -> true', () => {
	const {view, container} = fromString(
		'if (x)\n  | ${y}\nb | ${z}',
		{x: true, z: 3},
		{x: true, y: 'foo'},
	);

	expect(container.innerHTML).toBe('<!--if1--><!--(-->foo<!--)--><!--/if1--><b><!--(--><!--)--></b>');

	view.update({x: true, y: 'bar', z: 33});
	expect(container.innerHTML).toBe('<!--if1--><!--(-->bar<!--)--><!--/if1--><b><!--(-->33<!--)--></b>');

	view.update({});
	expect(container.innerHTML).toBe('<!--if1--><!--(--><!--)--><!--/if1--><b><!--(--><!--)--></b>');

	view.update({x: true, y: 'qux'});
	expect(container.innerHTML).toBe('qux<!--if1--><!--(--><!--)--><!--/if1--><b><!--(--><!--)--></b>');
});

it('iso / if / true -> false', () => {
	const {view, container} = fromString(
		'if (x)\n  | ${y}\nb | ${z}',
		{x: true, y: 'foo', z: 4},
		{},
	);

	expect(container.innerHTML).toBe('<!--if1--><!--/if1--><b><!--(--><!--)--></b>');

	view.update({x: true, y: 'bar', z: 44});
	expect(container.innerHTML).toBe('<!--if1-->bar<!--/if1--><b><!--(-->44<!--)--></b>');
});

it('iso / for / [] -> []', () => {
	const {view, container} = fromString(
		'for (i in x)\n  | ${i}\nb | ${y}',
		{y: 1},
		{y: 1},
	);

	expect(container.innerHTML).toBe('<!--for1--><!--/for1--><b><!--(-->1<!--)--></b>');

	view.update({x: ['a', 'b'], y: 11});
	expect(container.innerHTML).toBe('<!--for1-->ab<!--/for1--><b><!--(-->11<!--)--></b>');
});

it('iso / for / [] -> [a, b]', () => {
	const {view, container} = fromString(
		'for (i in x)\n  | ${i}\nb | ${y}',
		{y: 2},
		{x: ['a', 'b'], y: 22},
	);

	expect(container.innerHTML).toBe('<!--for1-->ab<!--/for1--><b><!--(-->22<!--)--></b>');

	view.update({x: ['c'], y: 222});
	expect(container.innerHTML).toBe('<!--for1-->c<!--/for1--><b><!--(-->222<!--)--></b>');

	view.update({x: [], y: 2222});
	expect(container.innerHTML).toBe('<!--for1--><!--/for1--><b><!--(-->2222<!--)--></b>');

	view.update({x: ['a']});
	expect(container.innerHTML).toBe('<!--for1-->a<!--/for1--><b><!--(--><!--)--></b>');
});

it('iso / for / [a, b] -> [a, b, c]', () => {
	const {view, container} = fromString(
		'for (i in x)\n  | ${i}\nb | ${y}',
		{x: ['a', 'b'], y: 3},
		{x: ['a', 'b', 'c'], y: 33},
	);

	expect(container.innerHTML).toBe('<!--for1--><!--(-->a<!--)--><!--(-->b<!--)-->c<!--/for1--><b><!--(-->33<!--)--></b>');

	view.update({x: ['a', 'b', 'c', 'd'], y: 333});
	expect(container.innerHTML).toBe('<!--for1--><!--(-->a<!--)--><!--(-->b<!--)-->cd<!--/for1--><b><!--(-->333<!--)--></b>');

	view.update({x: [], y: 3333});
	expect(container.innerHTML).toBe('<!--for1--><!--(--><!--)--><!--(--><!--)--><!--/for1--><b><!--(-->3333<!--)--></b>');

	view.update({x: ['a'], y: 3333});
	expect(container.innerHTML).toBe('<!--for1--><!--(--><!--)--><!--(--><!--)-->a<!--/for1--><b><!--(-->3333<!--)--></b>');
});

it('iso / for / [a, b, c] -> [a, b]', () => {
	const {view, container} = fromString(
		'for (i in x)\n  | ${i}\nb | ${y}',
		{x: ['a', 'b', 'c'], y: 4},
		{x: ['a', 'b'], y: 44},
	);

	expect(container.innerHTML).toBe('<!--for1--><!--(-->a<!--)--><!--(-->b<!--)--><!--/for1--><b><!--(-->44<!--)--></b>');

	view.update({x: ['a', 'x', 'c', 'd'], y: 444});
	expect(container.innerHTML).toBe('<!--for1--><!--(-->a<!--)--><!--(-->x<!--)-->cd<!--/for1--><b><!--(-->444<!--)--></b>');

	view.update({x: [], y: 4444});
	expect(container.innerHTML).toBe('<!--for1--><!--(--><!--)--><!--(--><!--)--><!--/for1--><b><!--(-->4444<!--)--></b>');

	view.update({x: ['a', 'b'], y: 4444});
	expect(container.innerHTML).toBe('<!--for1--><!--(--><!--)--><!--(--><!--)-->ab<!--/for1--><b><!--(-->4444<!--)--></b>');
});

it('iso / for / [a, b] -> []', () => {
	const {view, container} = fromString(
		'for (i in x)\n  | ${i}\nb | ${y}',
		{x: ['a', 'b'], y: 5},
		{x: [], y: 55},
	);

	expect(container.innerHTML).toBe('<!--for1--><!--/for1--><b><!--(-->55<!--)--></b>');

	view.update({x: ['a', 'b'], y: 555});
	expect(container.innerHTML).toBe('<!--for1-->ab<!--/for1--><b><!--(-->555<!--)--></b>');
});

it('iso / if + for', () => {
	const {view, container} = fromString(
		`
			form > if (x) > fieldset > for (i in y)
				if (i % 2) > b | Item: \${i}
				else > | Text: \${i}
			b | \${z}
		`,
		{x: true, y: [1, 2]},
		{x: true, y: [1, 2, 3]},
	);

	expect(container.innerHTML).toBe(
		'<form><!--if3--><fieldset><!--for2-->' +
		'<!--if1--><b>Item: <!--(-->1<!--)--></b><!--/if1-->' +
		'<!--if1-->Text: <!--(-->2<!--)--><!--/if1-->' +
		'<b>Item: 3</b>' +
		'<!--/for2--></fieldset><!--/if3--></form>' +
		'<b><!--(--><!--)--></b>'
	);

	view.update({x: true, y: [5], z: 'wow'});
	expect(container.innerHTML).toBe(
		'<form><!--if3--><fieldset><!--for2-->' +
		'<!--if1--><b>Item: <!--(-->5<!--)--></b><!--/if1-->' +
		'<!--if1--><!--(--><!--)--><!--/if1-->' +
		'<!--/for2--></fieldset><!--/if3--></form>' +
		'<b><!--(-->wow<!--)--></b>'
	);

	view.update({z: 'ok'});
	expect(container.innerHTML).toBe(
		'<form><!--if3--><!--/if3--></form>' +
		'<b><!--(-->ok<!--)--></b>'
	);

	view.update({x: true, y: [7, 8, 9]});
	expect(container.innerHTML).toBe(
		'<form><fieldset><!--for2-->' +
		'<!--if1--><b>Item: <!--(-->7<!--)--></b><!--/if1-->' +
		'<!--if1--><!--(--><!--)--><!--/if1-->' +
		'Text: 8' +
		'<b>Item: 9</b>' +
		'<!--/for2--></fieldset><!--if3--><!--/if3--></form>' +
		'<b><!--(--><!--)--></b>'
	);
});

it('iso / XFooSync', () => {
	const log = [];
	const __blocks__ = {
		XFoo: {
			template: 'h1 | ${attrs.name}',
			connectedCallback() { log.push('connected'); },
		},
	};
	const {view, container} = fromStringWithBlocks(
		'XFoo[name=${x}]',
		{__blocks__, x: 'fail'},
		{__blocks__, x: 'Exility'},
	);

	expect(log).toEqual(['connected']);
	expect(container.innerHTML).toBe('<h1><!--(-->Exility<!--)--></h1>');

	view.update({__blocks__, x: 'OK'});
	expect(container.innerHTML).toBe('<h1><!--(-->OK<!--)--></h1>');
});

it('iso / XFooAsync', async () => {
	const log = [];
	const __blocks__ = {
		XFoo: () => Promise.resolve({
			template: 'h1 | ${attrs.name}',
			connectedCallback() { log.push('connected'); },
		}),
	};
	const {view, container} = fromStringWithBlocks(
		'XFoo[name=${x}]',
		{__blocks__, x: 'fail'},
		{__blocks__, x: 'Exility'},
	);

	expect(log).toEqual([]);
	expect(container.innerHTML).toBe('<div data-block="XFoo" class="block-dummy block-dummy-loading"></div>');

	await pause();
	expect(log).toEqual(['connected']);

	expect(container.innerHTML).toBe('<h1>Exility</h1>');

	view.update({__blocks__, x: 'OK'});
	expect(container.innerHTML).toBe('<h1>OK</h1>');
});

it('iso / XFooSuper', () => {
	const XFoo = {template: 'div > ::children | Original:${attrs.value}'};
	const __blocks__ = {XFoo};
	const {view, container} = fromStringWithBlocks(
		'XFoo[value=${x}] > ::children\n  | ${y}:\n  ::super.children',
		{__blocks__, x: 1, y: 'fail'},
		{__blocks__, x: 2, y: 'OK'},
	);

	expect(container.innerHTML).toBe('<div><!--(-->OK<!--)-->:Original:<!--(-->2<!--)--></div>');

	view.update({__blocks__, x: 3, y: 'Yes'});

	expect(container.innerHTML).toBe('<div><!--(-->Yes<!--)-->:Original:<!--(-->3<!--)--></div>');
});

it('iso / page', async () => {
	const sandboxFactory = document.createElement('iframe');

	document.body.appendChild(sandboxFactory);
	await pause();

	const sandbox = sandboxFactory.contentDocument;

	sandbox.open();
	sandbox.write(createStringCompiler({
		scope: ['title', 'content', 'state'],
		metaComments: true,
	})(`
		!html
		html
			head > title[innerHTML=\${title}]
			body.is-\${state} | \${content}
	`)({stdlib})({
		state: 'backend',
		title: 'Initial',
		content: 'Ready',
	}));
	sandbox.close();

	expect(sandbox.title).toBe('Initial');
	expect(sandbox.body.className).toBe('is-backend');
	expect(sandbox.body.innerHTML).toBe('<!--(-->Ready<!--)-->');

	const view = createDOMCompiler({
		scope: ['title', 'content', 'state'],
		isomorphic: true,
	})(`
		!html
		html
			head > title[innerHTML=\${title}]
			body.is-\${state} | \${content}
	`)({stdlib, stddom})({
		state: 'client',
		title: 'Wow!',
		content: 'Norm',
	}, {
		isomorphic: sandbox
	});

	expect(sandbox.title).toBe('Wow!');
	expect(sandbox.body.className).toBe('is-client');
	expect(sandbox.body.innerHTML).toBe('<!--(-->Norm<!--)-->');

	view.update({
		state: 'interactive',
		title: 'Exility',
		content: 'Ultimate isomorphic page!',
	});

	expect(sandbox.title).toBe('Exility');
	expect(sandbox.body.className).toBe('is-interactive');
	expect(sandbox.body.innerHTML).toBe('<!--(-->Ultimate isomorphic page!<!--)-->');
});
