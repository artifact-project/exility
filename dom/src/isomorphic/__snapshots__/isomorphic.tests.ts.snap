// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__attrs__ 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CMP_SET_COMPILER = __STDDOM.CMP_SET_COMPILER;
	var __STDDOM_CMP_INIT = __STDDOM.CMP_INIT;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_CMP_CREATE = __STDDOM.CMP_CREATE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_CLONE_OBJECT = __STDLIB.CLONE_OBJECT;
	__STDDOM_CMP_SET_COMPILER(__COMPILER__);

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var val = __SCOPE__.val;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__;

		__STDDOM_CMP_INIT(__blocks__, ["Nested"]);

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __CMP_ATTRS = __STDLIB_CLONE_OBJECT((attrs));
		__CMP_ATTRS["baz"] = (val);
		var __x1 = __STDDOM_CMP_CREATE(__ctx, __blocks__,

			__frag,
			__this__,
			"Nested",
			__CMP_ATTRS,
			null,
			null
		);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				val = __NEWSCOPE__.val;
				__blocks__ = __NEWSCOPE__.__blocks__;
				var __CMP_ATTRS = __STDLIB_CLONE_OBJECT((attrs));
				__CMP_ATTRS["baz"] = (val);
				__x1.update(__CMP_ATTRS);
			}
		}
	}
}
`;

exports[`Btn -> Sync 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CMP_SET_COMPILER = __STDDOM.CMP_SET_COMPILER;
	var __STDDOM_CMP_INIT = __STDDOM.CMP_INIT;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_CMP_CREATE = __STDDOM.CMP_CREATE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	__STDDOM_CMP_SET_COMPILER(__COMPILER__);

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var val = __SCOPE__.val;
		var __blocks__ = __SCOPE__.__blocks__;
		var __this__ = __SCOPE__;

		__STDDOM_CMP_INIT(__blocks__, ["Btn"]);

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __CMP_ATTRS = {};
		__CMP_ATTRS["value"] = (val);
		var __x1 = __STDDOM_CMP_CREATE(__ctx, __blocks__,

			__frag,
			__this__,
			"Btn",
			__CMP_ATTRS,
			null,
			null
		);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				val = __NEWSCOPE__.val;
				__blocks__ = __NEWSCOPE__.__blocks__;
				var __CMP_ATTRS = {};
				__CMP_ATTRS["value"] = (val);
				__x1.update(__CMP_ATTRS);
			}
		}
	}
}
`;

exports[`CSS Module 1`] = `
function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var context = __SCOPE__.context;
		var __this__ = __SCOPE__.__this__;
		var __blocks__ = __SCOPE__.__blocks__;
		var __slots__ = __SCOPE__.__slots__;
		var __classNames__ = __SCOPE__.__classNames__;

		var __CLN = __STDLIB_CSS_MODULE(__classNames__);

		var __ctx = __STDDOM_CONTEXT();
		var __super__ = {
			"children": __SLOT_children
		};
		var __frag = __STDDOM_FRAGMENT();
		var __x1 = __STDDOM_LIVE_NODE(__frag, __ctx, 1, "div");
		__STDDOM_D_PROP(__ctx[1], "className", __CLN("alert " + (attrs.x) + " is-" + (attrs.y)))

		var __SLOT_children_used = false;
		function __SLOT_children() {
			__SLOT_children_used = true;
		}
		__STDLIB_SLOT(__slots__, __super__, "children", __SLOT_children, __x1);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				context = __NEWSCOPE__.context;
				__this__ = __NEWSCOPE__.__this__;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__slots__ = __NEWSCOPE__.__slots__;
				__classNames__ = __NEWSCOPE__.__classNames__;
				__CLN.upd(__classNames__);
				__STDDOM_D_PROP(__ctx[1], "className", __CLN("alert " + (attrs.x) + " is-" + (attrs.y)))
			}
		}
	}
`;

exports[`Icon (short) 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CMP_INLINE = __STDDOM.CMP_INLINE;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_LIVE_NODE = __STDDOM.LIVE_NODE;
	var __STDDOM_D_PROP = __STDDOM.D_PROP;
	var __STDDOM_CMP_CREATE_INLINE = __STDDOM.CMP_CREATE_INLINE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_TO_STRING = __STDLIB.TO_STRING;
	var __CMP_INLINE_STORE = {};
	__STDDOM_CMP_INLINE(__CMP_INLINE_STORE, "Icon", function (attrs) {
		var name = attrs.name;
		var __ctx = __STDDOM_CONTEXT();
		var __frag = __STDDOM_FRAGMENT();
		var __x2 = __STDDOM_LIVE_NODE(__frag, __ctx, 2, "i");
		__STDDOM_D_PROP(__ctx[2], "className", "icon-" + __STDLIB_TO_STRING(name))

		return {
			ctx: __ctx,
			frag: __frag,
			update: function (__NATTRS__) {
				attrs = __NATTRS__;
				name = __NATTRS__.name;
				__STDDOM_D_PROP(__ctx[2], "className", "icon-" + __STDLIB_TO_STRING(name))
			}
		};
	});

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;
		var __this__ = __SCOPE__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __CMP_ATTRS = {};
		__CMP_ATTRS["name"] = (x);
		var __x1 = __STDDOM_CMP_CREATE_INLINE(__CMP_INLINE_STORE,

			__frag,
			__this__,
			"Icon",
			__CMP_ATTRS,
			null,
			null
		);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				x = __NEWSCOPE__.x;
				var __CMP_ATTRS = {};
				__CMP_ATTRS["name"] = (x);
				__x1.update(__CMP_ATTRS);
			}
		}
	}
}
`;

exports[`Icon 1`] = `
function templateFactory(__DEPS__) {
	"use strict";
	var __STDDOM = __DEPS__.stddom;
	var __STDDOM_CMP_INLINE = __STDDOM.CMP_INLINE;
	var __STDDOM_CONTEXT = __STDDOM.CONTEXT;
	var __STDDOM_FRAGMENT = __STDDOM.FRAGMENT;
	var __STDDOM_LIVE_NODE = __STDDOM.LIVE_NODE;
	var __STDDOM_D_PROP = __STDDOM.D_PROP;
	var __STDDOM_CMP_CREATE_INLINE = __STDDOM.CMP_CREATE_INLINE;
	var __STDDOM_LIFECYCLE = __STDDOM.LIFECYCLE;
	var __STDLIB = __DEPS__.stdlib;
	var __STDLIB_TO_STRING = __STDLIB.TO_STRING;
	var __CMP_INLINE_STORE = {};
	__STDDOM_CMP_INLINE(__CMP_INLINE_STORE, "Icon", function (attrs) {
		var name = attrs.name;
		var __ctx = __STDDOM_CONTEXT();
		var __frag = __STDDOM_FRAGMENT();
		var __x2 = __STDDOM_LIVE_NODE(__frag, __ctx, 2, "i");
		__STDDOM_D_PROP(__ctx[2], "className", "icon-" + __STDLIB_TO_STRING(name))

		return {
			ctx: __ctx,
			frag: __frag,
			update: function (__NATTRS__) {
				attrs = __NATTRS__;
				name = __NATTRS__.name;
				__STDDOM_D_PROP(__ctx[2], "className", "icon-" + __STDLIB_TO_STRING(name))
			}
		};
	});

	return function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var x = __SCOPE__.x;
		var __this__ = __SCOPE__;

		var __ctx = __STDDOM_CONTEXT();

		var __frag = __STDDOM_FRAGMENT();
		var __CMP_ATTRS = {};
		__CMP_ATTRS["name"] = (x);
		var __x1 = __STDDOM_CMP_CREATE_INLINE(__CMP_INLINE_STORE,

			__frag,
			__this__,
			"Icon",
			__CMP_ATTRS,
			null,
			null
		);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				x = __NEWSCOPE__.x;
				var __CMP_ATTRS = {};
				__CMP_ATTRS["name"] = (x);
				__x1.update(__CMP_ATTRS);
			}
		}
	}
}
`;

exports[`Pseudo Elements: super 1`] = `
function compiledTemplate(__SCOPE__, __OPTIONS__) {
		__SCOPE__ = __SCOPE__ || {};
		var attrs = __SCOPE__.attrs;
		var context = __SCOPE__.context;
		var __this__ = __SCOPE__.__this__;
		var __blocks__ = __SCOPE__.__blocks__;
		var __slots__ = __SCOPE__.__slots__;
		var __classNames__ = __SCOPE__.__classNames__;

		var __ctx = __STDDOM_CONTEXT();
		var __super__ = {
			"children": __SLOT_children
		};
		var __frag = __STDDOM_FRAGMENT();
		var __x1 = __STDDOM_NODE(__frag, "div");

		var __SLOT_children_used = false;
		function __SLOT_children() {
			__SLOT_children_used = true;
			__STDDOM_TEXT(__x1, "Original:");
			__STDDOM_VALUE(__x1, __ctx, 2, attrs.value);
		}
		__STDLIB_SLOT(__slots__, __super__, "children", __SLOT_children, __x1);

		return {
			ctx: __ctx,
			frag: __frag,
			container: null,
			mountTo: function (container) {
				this.container = container;
				__frag.mountTo(container);
				__STDDOM_LIFECYCLE(__ctx, 'connectedCallback');
				return this;
			},
			update: function (__NEWSCOPE__) {
				__SCOPE__ = __NEWSCOPE__;
				attrs = __NEWSCOPE__.attrs;
				context = __NEWSCOPE__.context;
				__this__ = __NEWSCOPE__.__this__;
				__blocks__ = __NEWSCOPE__.__blocks__;
				__slots__ = __NEWSCOPE__.__slots__;
				__classNames__ = __NEWSCOPE__.__classNames__;

				if (__SLOT_children_used) {
					__STDDOM_UPD_VALUE(__ctx[2], attrs.value);
				}
			}
		}
	}
`;
